prow/tide/tide.go

```diff
@@ -930,26 +931,40 @@ func (pr *PullRequest) logFields() logrus.Fields {
 //
-// First, we try to get this value from the commits we got with the PR query.
-// Unfortunately the 'last' commit ordering is determined by author date
-// not commit date so if commits are reordered non-chronologically on the PR
-// branch the 'last' commit isn't necessarily the logically last commit.
-// We list multiple commits with the query to increase our chance of success,
-// but if we don't find the head commit we have to ask Github for it
-// specifically (this costs an API token).
+// There is no single way to use the GitHub Graphql API to get the status contexts
+// for the head commit of PR. There are 2 ways that each only work under certain
+// conditions and sometimes neither of those will work and we must fall back to
+// the REST API. Here is our process:
+//
+// First, we try to get the statuses via the `headRef` field. This works most PRs,
+// only PRs with a deleted head ref will need to continue on.
+//
+// If the head ref has been deleted, we can still get status contexts by looking
+// at the list of commits for the PR. Unfortunately the 'last' commit ordering
+// is determined by author date not commit date so if commits are reordered
+// non-chronologically on the PR branch, the 'last' commit isn't necessarily the
+// logically last commit. Most PRs won't have reorder commits so in most cases
+// we can just use the statuses from this 'last' commit.
+//
+// If the worst case occurs and neither of the above cases is suitable we have
+// to use the REST API to get the head commit statuses. This costs an extra API
+// token every sync loop.
+//
+// Here are some issues on GitHub's support forum that describe why this gross
+// work around is necessary.
+// https://platform.github.community/t/some-prs-are-missing-head-refs/4586
+// https://platform.github.community/t/github-commits-returned-in-the-incorrect-order-how-to-get-the-head-commit-for-statuses/4130
 func headContexts(log *logrus.Entry, ghc githubClient, pr *PullRequest) ([]Context, error) {
-	for _, node := range pr.Commits.Nodes {
-		if node.Commit.OID == pr.HeadRefOID {
-			return node.Commit.Status.Contexts, nil
-		}
+	if contexts, ok := headContextsNoCost(pr); ok {
+		return contexts, nil
 	}
-	// We didn't get the head commit from the query (the commits must not be
+
+	// We didn't get the head commit from the query (the PR's commits must not be
 	// logically ordered) so we need to specifically ask Github for the status
 	// and coerce it to a graphql type.
 	org := string(pr.Repository.Owner.Login)
 	repo := string(pr.Repository.Name)
-	// Log this event so we can tune the number of commits we list to minimize this.
-	log.Warnf("'last' %d commits didn't contain logical last commit. Querying Github...", len(pr.Commits.Nodes))
+	log.Warnf("HeadRef was missing and 'last' commit is not the logical last commit. Querying Github...")
 	combined, err := ghc.GetCombinedStatus(org, repo, string(pr.HeadRefOID))
```

Comment:
as a follow up at some point when we get prometheus metrics from tide (like pool sizes, success rates etc) we should also record how often we need this